### PR TITLE
Frontend: Configure karma, travis & docker to run unit tests along with writing tests for analytics, auth & create challenge controllers

### PR DESCRIPTION
**Pull request by Sanji515**
**Deliverables:**
- [x] Configured packages `Karma` and `Jasmine`
- [x] Changed `angular-mocksv1.6.1` to `angular-mocksv1.7.0`
- [x] Changed from `Mocha` syntax to `Jasmine` syntax and added remaining tests for `authCtrl.js`
- [x] Added tests for `analyticsCtrl.js`
- [x] Added tests for `challengeCreate.js`

**To run the Frontend Unit Tests:**

Run the following:
- `docker-compose up --build` (to install the required dependencies )
- `docker-compose exec nodejs bash` (go inside the nodejs container)
- `karma start` (run the tests)

### CODE REVIEW

**Comment by Sanji515** [comment]
@RishabhJain2018 @galipremsagar Please review.

File: frontend/src/js/controllers/challengeCreate.js [hunk 1]
```diff
@@ -39,8 +43,13 @@
                     
                 }
                 if (vm.input_file) {
+                    vm.isExistLoader = true;
+                    vm.loaderTitle = '';
+                    vm.loaderContainer = angular.element('.exist-team-card');
+                    vm.startLoader('create challenge');
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCreate.js]: Why was this added?

File: frontend/tests/controllers-test/analyticsCtrl.test.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
+'use strict';
```
Comment (galipremsagar) [inline frontend/tests/controllers-test/analyticsCtrl.test.js]: Can you indent this file? Would be ideal to have all files you commit to have them indented. At least the ones(files or part of chunks of code) being newly added by you.

File: frontend/tests/controllers-test/authCtrl.test.js [hunk 3]
```diff
@@ -191,23 +192,111 @@ describe('Unit Tests for auth controller', function() {
...
+                    color = "darkgreen";
+                } else if (passed == 5) {
+                    strength = "Very Strong";
+                    color = "darkgreen";
+                }
+                return [strength, color];
+            }
+        });
+
+        it('check Weak password', function() {
```
Comment (galipremsagar) [inline frontend/tests/controllers-test/authCtrl.test.js]: Club all the password strengths into one unit test, and maintain a list of objects with password, expectedStrength & expectedColor. Iterate over that list and do a similar type of validation inside for loop. Let's avoid duplication here, In this way adding more password strength checks will be much easier in future.

File: package.json [hunk 4]
```diff
@@ -9,7 +9,7 @@
   "author": "",
   "license": "ISC",
   "dependencies": {
-    "angular-mocks": "^1.6.1",
+    "angular-mocks": "^1.7.0",
```
Comment (galipremsagar) [inline package.json]: Any specific reason to choose 1.7.0 when we have 1.7.8?

File: frontend/src/js/controllers/challengeCreate.js (hunk 1 above)
Comment (Sanji515) [reply frontend/src/js/controllers/challengeCreate.js]: @galipremsagar  done. Removed the unnecessary lines as I've already initialized them at the start

File: frontend/tests/controllers-test/analyticsCtrl.test.js (hunk 2 above)
Comment (Sanji515) [reply frontend/tests/controllers-test/analyticsCtrl.test.js]: @galipremsagar It's already indented correctly but not showing same as in my editor.

File: frontend/tests/controllers-test/authCtrl.test.js (hunk 3 above)
Comment (Sanji515) [reply frontend/tests/controllers-test/authCtrl.test.js]: @galipremsagar done. Please have a look if it's in the same way as you wanted.

File: package.json (hunk 4 above)
Comment (Sanji515) [reply package.json]: @galipremsagar yes, we cannot use `angular 1.7.0` with `angular-mocks 1.7.8`, the version of both of them should be matched.

**Comment by Sanji515** [comment]
@galipremsagar Maybe it's because of the version of `angular-mocks`, please change it to `version 1.7.0`

File: package.json [hunk 5]
```diff
@@ -38,5 +38,11 @@
   "repository": {
     "type": "git",
     "url": "https://github.com/Cloud-CV/EvalAI.git"
+  },
+  "devDependencies": {
+    "jasmine-core": "^3.4.0",
```
Comment (galipremsagar) [inline package.json]:
```suggestion
    "jasmine-core": "^3.4.0",
    "angular-mocks": "^1.7.0",
```

Lets keep angular-mocks inside dev-dependencies rather than dependency.

File: package.json (hunk 4 above)
Comment (galipremsagar) [inline package.json]:
```suggestion
```

**Comment by galipremsagar** [comment]
> @galipremsagar It's because of the version of `angular-mocks`, please change it to `version 1.7.0`

Okay, works now.

Screenshot 2019-04-24 at 1 21 38 PM

**Comment by Sanji515** [comment]
@galipremsagar applied the suggested changes, please review.
thanks!

File: frontend/tests/controllers-test/authCtrl.test.js (hunk 3 above)
Comment (galipremsagar) [reply frontend/tests/controllers-test/authCtrl.test.js]: Hey, I meant something different, I've made those changes. You will find a patch file in this zip. Please apply this path with this entire file and push the changes.

[Merged_all_code_duplication_into_one_single_function_which_does_the_same___Co-authored-by_.patch.zip](https://github.com/Cloud-CV/EvalAI/files/3114908/Merged_all_code_duplication_into_one_single_function_which_does_the_same___Co-authored-by_.patch.zip)

**Comment by Sanji515** [comment]
thanks, @galipremsagar  applied the patch, please review.

File: frontend/tests/controllers-test/authCtrl.test.js (hunk 3 above)
Comment (Sanji515) [reply frontend/tests/controllers-test/authCtrl.test.js]: done.

**Comment by Sanji515** [comment]
@RishabhJain2018 Please review

**Comment by RishabhJain2018** [comment]
@Sanji515 Can you please update the readme for how to run the tests?

**Comment by RishabhJain2018** [comment]
Thanks, @galipremsagar for the review. I'll merge it soon.

**Comment by RishabhJain2018** [comment]
Sorry, @Sanji515 edit: Please update the PR description.

**Comment by Sanji515** [comment]
@RishabhJain2018 Updated the PR description, please check.

**Comment by RishabhJain2018** [comment]
@Sanji515 I tried running it locally but it doesn't work. Can you please do a fresh build and then check it.
cc: @galipremsagar

**Comment by Sanji515** [comment]
@RishabhJain2018 I've checked on a fresh build, it's working on my side. Can you please show the error?

**Comment by Sanji515** [comment]
@RishabhJain2018 Is it still failing on your side?

**Comment by Sanji515** [comment]
@RishabhJain2018 Please take a look.

**Comment by Sanji515** [comment]
Hi @RishabhJain2018, is it still failing on your local?

**Comment by RishabhJain2018** [comment]
@galipremsagar Can you please test if the tests work for you inside the docker container?

**Comment by galipremsagar** [comment]
> @galipremsagar Can you please test if the tests work for you inside the docker container?

@RishabhJain2018 Correct me if I did something wrong, else sanjeev will need to fix `CHROME_BIN` on docker container:
1. Logged into evalai_nodejs image
2. Ran this command: `/usr/local/bin/node ./node_modules/karma/bin/karma start`

This was the error I'm getting:
Screenshot 2019-05-14 at 6 16 57 AM

In case these steps are valid @Sanji515  please look into it

**Comment by Sanji515** [comment]
@RishabhJain2018 @galipremsagar I've fixed that issue, please check

File: docker/dev/nodejs/Dockerfile [hunk 6]
```diff
@@ -11,6 +11,7 @@ ADD ./karma.conf.js /code
 
 # Install Prerequisites
 RUN npm install -g bower gulp
+RUN apt-get install gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget
```
Comment (galipremsagar) [inline docker/dev/nodejs/Dockerfile]: This command is causing docker build to fail. 
Screenshot 2019-05-14 at 7 26 02 PM

**Comment by Sanji515** [comment]
@RishabhJain2018 @galipremsagar added backports to the sources-list which will enable `apt` to run from the new sources, hence will be able to locate the packages. It's working on my side, please review

**Comment by galipremsagar** [comment]
Works fine for me now, I'm able to run unit tests now inside the container. Though would suggest @RishabhJain2018 too to test it out. Something to watch out for was I was having some timeout issues for package installs.

File: pytest.ini [hunk 7]
```diff
@@ -1,4 +1,4 @@
 [pytest]
 python_files=test*.py
 DJANGO_SETTINGS_MODULE=settings.test
-norecursedirs=env venv
+norecursedirs=env venv node_modules bower_components
```
Comment (RishabhJain2018) [inline pytest.ini]: Please remove this change. It doesn't correspond to this PR.

File: setup.cfg [hunk 8]
```diff
@@ -1,6 +1,6 @@
 [flake8]
 max-line-length = 120
-exclude = .git,*/migrations/*,*/static/CACHE/*,docs/,env/,fabfile/
+exclude = .git,*/migrations/*,*/static/CACHE/*,docs/,env/,fabfile/,node_modules/,bower_components/
```
Comment (RishabhJain2018) [inline setup.cfg]: Please remove this change. It doesn't correspond to this PR.

File: .travis.yml [hunk 9]
```diff
@@ -28,11 +33,15 @@ before_script:
...
 - sudo rm -f /etc/boto.cfg
+- export CHROME_BIN=chromium-browser
+- export DISPLAY=:99.0
+- sh -e /etc/init.d/xvfb start
 install:
 - pip install -r requirements/dev.txt
 - pip install awscli==1.16.57 coveralls
 script:
 - flake8 ./
+- node_modules/karma/bin/karma start karma.conf.js --single-run
```
Comment (RishabhJain2018) [inline .travis.yml]: Can we please install `karma-cli` in order to avoid this long command. For reference: http://karma-runner.github.io/4.0/intro/installation.html

File: .coveragerc [hunk 10]
```diff
@@ -14,6 +14,8 @@ omit =
     settings/dev.py
     settings/prod.py
     scripts/seed.py
+    node_modules/
```
Comment (RishabhJain2018) [inline .coveragerc]: Please remove this change as this doesn't correspond to this PR.

File: .coveragerc [hunk 11]
```diff
@@ -14,6 +14,8 @@ omit =
...
+    bower_components/
```
Comment (RishabhJain2018) [inline .coveragerc]: Please remove this change as this doesn't correspond to this PR.

File: .travis.yml (hunk 9 above)
Comment (RishabhJain2018) [reply .travis.yml]: Please do the same for running tests in nodejs docker container.

**Comment by Sanji515** [comment]
@RishabhJain2018 Please review

File: frontend/tests/controllers-test/analyticsCtrl.test.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+describe('Unit Tests for analytics controller', function() {
```
[the remark this answers is not in the document]
Comment (Sanji515) [reply frontend/tests/controllers-test/analyticsCtrl.test.js]: done

File: docker/dev/nodejs/Dockerfile [hunk 13]
```diff
@@ -12,10 +12,23 @@ ADD ./karma.conf.js /code
 # Install Prerequisites
 RUN npm install -g bower gulp
 
+# This enables `apt` to run from the new sources
+RUN printf "deb http://archive.debian.org/debian/ jessie main\ndeb-src http://archive.debian.org/debian/ jessie main\ndeb http://security.debian.org jessie/updates main\ndeb-src http://security.debian.org jessie/updates main" > /etc/apt/sources.list
```
[the remark this answers is not in the document]
Comment (Sanji515) [reply docker/dev/nodejs/Dockerfile]: done
